src/frontend/src/optimizer/rule/pull_up_correlated_predicate.rs

```diff
@@ -0,0 +1,136 @@
+// Copyright 2022 Singularity Data
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use itertools::{Either, Itertools};
+
+use super::super::plan_node::*;
+use super::{BoxedRule, Rule};
+use crate::expr::{CorrelatedInputRef, Expr, ExprImpl, ExprRewriter, InputRef};
+use crate::optimizer::PlanRef;
+use crate::utils::Condition;
+
+/// This rule is for pattern: Apply->Project->Filter.
+///
+/// To unnest, we just pull predicates contain correlated variables in Filter into Apply, and
+/// convert it into corresponding type of Join.
+pub struct PullUpCorrelatedPredicate {}
+impl Rule for PullUpCorrelatedPredicate {
+    fn apply(&self, plan: PlanRef) -> Option<PlanRef> {
+        let apply = plan.as_logical_apply()?;
+
+        let right = apply.right();
+        let project = right.as_logical_project()?;
+        let (mut exprs, mut expr_alias, _) = project.clone().decompose();
+        let begin = exprs.len();
+
+        let input = project.input();
+        let filter = input.as_logical_filter()?;
+
+        // Split predicates in LogicalFilter into correlated expressions and uncorrelated
+        // expressions.
+        let (cor_exprs, uncor_exprs) =
+            filter
+                .predicate()
+                .clone()
+                .into_iter()
+                .partition_map(|expr| {
+                    let mut rewriter = Rewriter {
+                        has_correlated_input_ref: false,
+                        input_refs: vec![],
+                        index: exprs.len() + apply.left().schema().fields().len(),
+                    };
+                    let rewritten_expr = rewriter.rewrite_expr(expr.clone());
+                    if rewriter.has_correlated_input_ref {
+                        // Append input_refs in expression which contains correlated_input_ref to
+                        // `exprs` used to construct new LogicalProject.
```

Comment:
```suggestion
                        // Append `InputRef`s in the predicate expression to be pulled to the project, so that they are accessible by the expression after it is pulled.
```